src/session/content/message_row/mod.rs

```diff
@@ -1,224 +1,115 @@
-            }
-        }
-
-        fn dispose(&self, _obj: &Self::Type) {
-            if let Some(avatar) = self.avatar.borrow().as_ref() {
+    impl MessageRow {
+        fn dispose(&self, _obj: &super::MessageRow) {
+            if let Some(avatar) = self.avatar.take() {
                 avatar.unparent();
             }
 
-            if let Some(content) = self.content.borrow().as_ref() {
+            if let Some(content) = self.content.take() {
                 content.unparent();
             }
         }
-    }
-
-    impl WidgetImpl for MessageRow {
-        fn measure(
-            &self,
-            _widget: &Self::Type,
-            orientation: gtk::Orientation,
-            for_size: i32,
-        ) -> (i32, i32, i32, i32) {
-            let (mut minimum, mut natural, mut minimum_baseline, mut natural_baseline) =
-                (0, 0, -1, -1);
-            let mut remaining_for_size = for_size;
-
-            if let Some(avatar) = self.avatar.borrow().as_ref() {
-                let (avatar_minimum, avatar_natural, _, _) = avatar.measure(orientation, for_size);
-                minimum = avatar_minimum;
-                natural = avatar_natural;
-
-                if let gtk::Orientation::Horizontal = orientation {
-                    minimum += SPACING;
-                    natural += SPACING;
-                } else if remaining_for_size != -1 {
-                    let (_, avatar_natural_opposite, _, _) =
-                        avatar.measure(gtk::Orientation::Horizontal, avatar_natural);
-                    remaining_for_size -= avatar_natural_opposite + SPACING;
-                }
-            }
-
-            if let Some(content) = self.content.borrow().as_ref() {
-                let (
-                    content_minimum,
-                    content_natural,
-                    content_minimum_baseline,
-                    content_natural_baseline,
-                ) = {
-                    let (minimum, mut natural, minimum_baseline, natural_baseline) =
-                        content.measure(orientation, remaining_for_size);
-                    let (_, default_natural, _, _) = content.measure(orientation, -1);
-
-                    // Always prefer default natural size
-                    if natural > default_natural && default_natural >= minimum {
-                        natural = default_natural
-                    }
-
-                    (minimum, natural, minimum_baseline, natural_baseline)
-                };
 
-                minimum_baseline = content_minimum_baseline;
-                natural_baseline = content_natural_baseline;
-
-                if let gtk::Orientation::Horizontal = orientation {
-                    minimum += content_minimum;
-                    natural += content_natural;
-                } else {
-                    minimum = minimum.max(content_minimum);
-                    natural = natural.max(content_natural);
+        #[widget_action]
+        fn delete(&self) {
```

Comment:
This action should not always be active. We need a general function that can disable/activate the actions that the related messages can actually activate. Also the message menu should not display items whose actions are disabled.